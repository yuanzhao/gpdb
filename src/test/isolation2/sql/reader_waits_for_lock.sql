-- This test validates, reader and writer both wait if lock is not
-- available. There used to be a bug where reader didn't wait even if lock was
-- held by some other session.

-- setup
5: create table reader_waits_for_lock_table(a int, b int) distributed by (a);
5: insert into reader_waits_for_lock_table select 1, 1;
-- save session id
5: CREATE TABLE reader_waits_for_lock_table_sessionid(a, setting) AS SELECT 1, setting::int FROM pg_settings WHERE name = 'gp_session_id' distributed by (a);
2U: BEGIN;
2U: LOCK reader_waits_for_lock_table IN ACCESS EXCLUSIVE MODE;
-- creates reader and writer gang
5&: SELECT t1.* FROM reader_waits_for_lock_table t1 INNER JOIN reader_waits_for_lock_table t2 ON t1.b = t2.b;
-- all processes in the session 5 should be blocked
2U: SELECT count(case when not waiting then 1 end), count(*) > 0 FROM pg_stat_activity where sess_id = (SELECT setting FROM reader_waits_for_lock_table_sessionid);
2U: COMMIT;
5<:
