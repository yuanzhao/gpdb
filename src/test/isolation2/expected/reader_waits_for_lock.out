-- This test validates, reader and writer both wait if lock is not
-- available. There used to be a bug where reader didn't wait even if lock was
-- held by some other session.

-- setup
5: create table reader_waits_for_lock_table(a int, b int) distributed by (a);
CREATE
5: insert into reader_waits_for_lock_table select 1, 1;
INSERT 1
-- save session id
5: CREATE TABLE reader_waits_for_lock_table_sessionid(a, setting) AS SELECT 1, setting::int FROM pg_settings WHERE name = 'gp_session_id' distributed by (a);
CREATE 1
2U: BEGIN;
BEGIN
2U: LOCK reader_waits_for_lock_table IN ACCESS EXCLUSIVE MODE;
LOCK
-- creates reader and writer gang
5&: SELECT t1.* FROM reader_waits_for_lock_table t1 INNER JOIN reader_waits_for_lock_table t2 ON t1.b = t2.b;  <waiting ...>
-- all processes in the session 5 should be blocked
2U: SELECT count(case when not waiting then 1 end), count(*) > 0 FROM pg_stat_activity where sess_id = (SELECT setting FROM reader_waits_for_lock_table_sessionid);
count|?column?
-----+--------
0    |t       
(1 row)
2U: COMMIT;
COMMIT
5<:  <... completed>
a|b
-+-
1|1
(1 row)
